mani_skill/envs/tasks/tabletop/roll_ball.py

```diff
@@ -0,0 +1,179 @@
+from mani_skill.utils.scene_builder.table import TableSceneBuilder
+from mani_skill.utils.structs.pose import Pose
+from mani_skill.utils.structs.types import Array, GPUMemoryConfig, SimConfig
+from transforms3d.euler import euler2quat
+
+@register_env("RollBall-v1", max_episode_steps=80)
+class RollBallEnv(BaseEnv):
+    """
+    Task Description
+    ----------------
+    A simple task where the objective is to push and roll a ball to a goal region at the other end of the table
+
+    Randomizations
+    --------------
+    - the ball's xy position is randomized on top of a table in the region [0.2, 0.5] x [-0.4, 0.7]. It is placed flat on the table
+    - the target goal region is marked by a red/white circular target. The position of the target is randomized on top of a table in the region [-0.4, -0.7] x [0.2, -0.9]
+
+    Success Conditions
+    ------------------
+    - the ball's xy position is within goal_radius (default 0.1) of the target's xy position by euclidean distance.
+    """
+
+    SUPPORTED_ROBOTS = ["panda"]
+
+    agent: Union[Panda]
+
+    goal_radius: float = 0.1 # radius of the goal region
+    ball_radius: float = 0.035 # radius of the ball
+    reached_status: torch.Tensor
+
+
+
+    def __init__(self, *args, robot_uids="panda", robot_init_qpos_noise=0.02, **kwargs):
+        self.robot_init_qpos_noise = robot_init_qpos_noise
+        self.reached_status = torch.zeros(kwargs["num_envs"], dtype=torch.float32)
+        super().__init__(*args, robot_uids=robot_uids, **kwargs)
+
+    @property
+    def _default_sim_config(self):
+        return SimConfig(
+            gpu_memory_cfg=GPUMemoryConfig(
+                found_lost_pairs_capacity=2**25, max_rigid_patch_count=2**18
+            )
+        )
+
+    @property
+    def _default_sensor_configs(self):
+        pose = sapien_utils.look_at(eye=[-0.1, 0.9, 0.3], target=[0.0, 0.0, 0.0])
+        return [CameraConfig("base_camera", pose, 128, 128, np.pi / 2, 0.01, 100)]
+    
+    @property
+    def _default_human_render_camera_configs(self):
+        pose = sapien_utils.look_at([-0.6, 1.3, 0.8], [0.0, 0.13, 0.0])
+        return CameraConfig("render_camera", pose, 512, 512, 1, 0.01, 100)
+
+
+    def _load_scene(self, options: dict):
+        self.table_scene = TableSceneBuilder(
+            self, robot_init_qpos_noise=self.robot_init_qpos_noise
+        )
+        self.table_scene.build()
+
+        self.ball = actors.build_sphere(
+            self.scene,
+            radius=self.ball_radius,
+            color=[0, 0.2, 0.8, 1],
+            name="ball",
+        )
+        
+
+        self.goal_region = actors.build_red_white_target(
+            self.scene,
+            radius=self.goal_radius,
+            thickness=1e-5,
+            name="goal_region",
+            add_collision=False,
+            body_type="kinematic",
+        )
+
+    def _initialize_episode(self, env_idx: torch.Tensor, options: dict):
+        if self.reached_status.device != self.device:
```

Comment:
i don't understand why this needs to be done. How come reached status bool needs to be delayed one step for the reward function? Why not compute it in the reward function based on the tcp position to ball distance instantaneously?

Reply:
The "reached_status" keeps the track of envs that manged to reach behind the ball, post which the reward function returns the reward in the next stage, i.e rolling the ball by hitting.

I don't understand what you mean by delayed ? can you explain ?

I need to reset it every time the a env gets a reset. so i kept it there. (also initializing in device is tricky because when i went through the source code the device implementation is not perfect so, I tried to do a future proof implementation)

line 165 to 172.
`        if num_envs > 1 or sim_backend == "gpu":
            if not physx.is_gpu_enabled():
                physx.enable_gpu()
            self.device = torch.device(
                "cuda"
            )  # TODO (stao): fix this for multi process support
        else:
            self.device = torch.device("cpu")
`

Reply:
you can safely use self.device. Each env object should be associated with one device.

for reached_status, I notice that you update this value after all the reward is done computing, hence it is delayed by one step. This does not make much sense (and it might slow RL down).

Reply:
Yeah that is a problem, i just committed a fix.